typespec-extension/src/code-model-builder.ts

```diff
@@ -2228,81 +2037,37 @@ export class CodeModelBuilder {
+        valueType: type.additionalProperties,
+      };
+      const parentSchema = this.processDictionarySchemaFromSdkType(sdkDictType, "Record");
+      objectSchema.parents = objectSchema.parents ? objectSchema.parents : new Relations();
       objectSchema.parents.immediate.push(parentSchema);
       pushDistinct(objectSchema.parents.all, parentSchema);
-    }
-
-    // value of the discriminator property
-    if (objectSchema.parents) {
-      const parentWithDiscriminator = objectSchema.parents.all.find(
-        (it) => it instanceof ObjectSchema && it.discriminator,
-      );
-      if (parentWithDiscriminator) {
-        discriminatorPropertyName = (
-          (parentWithDiscriminator as ObjectSchema).discriminator as discriminatorTypeWithPropertyName
-        ).propertyName;
-
-        const discriminatorProperty = Array.from(type.properties.values()).find(
-          (it) =>
-            it.name === discriminatorPropertyName &&
-            (it.type.kind === "String" || it.type.kind === "EnumMember" || it.type.kind === "UnionVariant"),
-        );
-        if (discriminatorProperty) {
-          if (discriminatorProperty.type.kind === "String") {
-            // value as StringLiteral
-            objectSchema.discriminatorValue = discriminatorProperty.type.value;
-          } else if (discriminatorProperty.type.kind === "EnumMember") {
-            // value as EnumMember
-            // lint requires value be string
-            objectSchema.discriminatorValue =
-              (discriminatorProperty.type.value as string) ?? discriminatorProperty.type.name;
-          } else if (discriminatorProperty.type.kind === "UnionVariant") {
-            // value as UnionVariant
-            objectSchema.discriminatorValue =
-              ((discriminatorProperty.type.type as StringLiteral).value as string) ?? discriminatorProperty.type.name;
-          }
-        } else {
-          // it is possible that the property is Union, e.g. 'kind: "type1" | "type2"'; but such Type appears not to be a concrete model.
-
-          // fallback to name of the Model
-          objectSchema.discriminatorValue = name;
-        }
-      }
+      objectSchema.discriminatorValue = type.discriminatorValue;
     }
 
     // properties
-    for (const prop of type.properties.values()) {
-      if (
-        prop.name === discriminatorPropertyName || // skip the discriminator property
-        isNeverType(prop.type) || // skip property of type "never"
-        !isPayloadProperty(this.program, prop)
-      ) {
-        continue;
+    for (const prop of type.properties) {
+      if (prop.kind === "property" && prop.discriminator === false) {
+        objectSchema.addProperty(this.processModelPropertyFromSdkType(prop));
       }
-      objectSchema.addProperty(this.processModelProperty(prop));
     }
 
-    // process all children
-    type.derivedModels?.filter(modelContainsDerivedModel).forEach((it) => this.processSchema(it, this.getName(it)));
```

Comment:
The removal of this could cause certain behavior difference. Though I am not yet sure whether the difference be expected.

Reply:
It basically about the whether a `model B extends A` get generated as public or not, when `B` is not used in any op.

Not a blocker. It likely going to have effect when you do ussage.